cpp/src/arrow/io/buffered.cc

```diff
@@ -178,5 +217,215 @@ Status BufferedOutputStream::Flush() { return impl_->Flush(); }
 
 std::shared_ptr<OutputStream> BufferedOutputStream::raw() const { return impl_->raw(); }
 
+// ----------------------------------------------------------------------
+// BufferedInputStream implementation
+
+class BufferedInputStream::BufferedInputStreamImpl : public BufferedBase {
+ public:
+  BufferedInputStreamImpl(std::shared_ptr<InputStream> raw, MemoryPool* pool)
+      : BufferedBase(pool),
+        raw_(std::move(raw)),
+        supports_zero_copy_(raw_->supports_zero_copy()),
+        bytes_buffered_(0) {}
+
+  ~BufferedInputStreamImpl() { DCHECK(Close().ok()); }
+
+  Status Close() {
+    std::lock_guard<std::mutex> guard(lock_);
+    if (is_open_) {
+      is_open_ = false;
+      return raw_->Close();
+    }
+    return Status::OK();
+  }
+
+  Status Tell(int64_t* position) const {
+    std::lock_guard<std::mutex> guard(lock_);
+    if (raw_pos_ == -1) {
+      RETURN_NOT_OK(raw_->Tell(&raw_pos_));
+      DCHECK_GE(raw_pos_, 0);
+    }
+    // Shift by bytes_buffered to return semantic stream position
+    *position = raw_pos_ - bytes_buffered_;
+    return Status::OK();
+  }
+
+  Status SetBufferSize(int64_t new_buffer_size) {
+    std::lock_guard<std::mutex> guard(lock_);
+    DCHECK_GT(new_buffer_size, 0);
+    if ((buffer_pos_ + bytes_buffered_) >= new_buffer_size) {
+      return Status::Invalid("Cannot shrink read buffer if buffered data remains");
+    }
+    return ResizeBuffer(new_buffer_size);
+  }
+
+  util::string_view Peek(int64_t nbytes) const {
+    int64_t peek_size = std::min(nbytes, bytes_buffered_);
+    return util::string_view(reinterpret_cast<const char*>(buffer_data_ + buffer_pos_),
```

Comment:
`buffer_data` may be null here?

Reply:
No; buffer_data_ is initialized when `Create` is called